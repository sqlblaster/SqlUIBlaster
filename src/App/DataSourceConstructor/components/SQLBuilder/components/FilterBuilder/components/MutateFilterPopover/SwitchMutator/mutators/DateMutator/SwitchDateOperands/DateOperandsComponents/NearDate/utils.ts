import { PluralDateComponent } from './plural-date-components';

export const getThisDateComponentNotation = (
  dateComponent: PluralDateComponent
) => {
  const singularForm = dateComponent.substr(0, dateComponent.length - 1);

  const thisDateNotation =
    singularForm === 'Day' ? 'Today' : `This ${singularForm}`;

  return thisDateNotation.toLowerCase();
};
