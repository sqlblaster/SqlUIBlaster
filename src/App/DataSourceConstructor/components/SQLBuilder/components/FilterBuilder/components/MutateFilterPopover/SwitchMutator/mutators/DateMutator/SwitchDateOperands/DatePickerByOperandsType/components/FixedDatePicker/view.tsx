import { InputAdornment, Typography, withStyles } from '@material-ui/core';
import AccessTimeIcon from '@material-ui/icons/AccessTime';
import { DatePicker, TimePicker } from 'material-ui-pickers';
import * as React from 'react';
import { RemoveButton } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { FixedDatePickerViewProps } from './props';
import { fixedDatePickerStyles } from './styles';

export const addATimeText = 'Add a time';

export const FixedDatePicker = withStyles(fixedDatePickerStyles)(
  class extends React.Component<FixedDatePickerViewProps> {
    public render() {
      const {
        date,
        timeEnabled,
        limits,
        classes,
        handleDateChange,
        handleAddTimeClick,
        handleRemoveTimeClick
      } = this.props;

      return (
        <div className={classes.root}>
          <DatePicker
            value={date}
            onChange={handleDateChange}
            format='DD/MM/YYYY'
            margin='none'
            variant='outlined'
            style={{ width: '100%', marginTop: '10px' }}
            InputProps={{
              inputProps: {
                style: { padding: '7px 14px', fontSize: '13px' }
              }
            }}
            {...(limits ? limits : {})}
          />
          {!timeEnabled && (
            <label className={classes['add-time']} onClick={handleAddTimeClick}>
              <AccessTimeIcon
                style={{
                  marginRight: '5px',
                  fontSize: '16px',
                  fontWeight: 'bold'
                }}
              />
              <Typography color='inherit'>{addATimeText}</Typography>
            </label>
          )}
          {timeEnabled && (
            <TimePicker
              value={date}
              onChange={handleDateChange}
              margin='none'
              variant='outlined'
              style={{ width: '100%', marginTop: '10px' }}
              InputProps={{
                inputProps: {
                  style: { padding: '7px 14px', fontSize: '13px' }
                },
                endAdornment: (
                  <InputAdornment position='end'>
                    <RemoveButton onClick={handleRemoveTimeClick} />
                  </InputAdornment>
                )
              }}
            />
          )}
        </div>
      );
    }
  }
);
