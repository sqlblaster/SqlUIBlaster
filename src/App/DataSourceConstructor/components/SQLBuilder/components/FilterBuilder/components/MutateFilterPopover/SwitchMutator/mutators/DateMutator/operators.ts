enum DateOperatorEnum {
  'Previous',
  'Next',
  'Before',
  'After',
  'On',
  'Between',
  'Is empty',
  'Is not empty'
}

export type DateOperator = keyof typeof DateOperatorEnum;

export const dateOperators: DateOperator[] = [
  'Previous',
  'Next',
  'Before',
  'After',
  'On',
  'Between',
  'Is empty',
  'Is not empty'
];
