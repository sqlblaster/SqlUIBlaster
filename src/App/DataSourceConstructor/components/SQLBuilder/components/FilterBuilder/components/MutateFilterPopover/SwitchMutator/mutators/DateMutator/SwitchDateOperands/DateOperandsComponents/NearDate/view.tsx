import { Checkbox, FormControlLabel, TextField, withStyles } from '@material-ui/core';
import * as React from 'react';
import { CheckedSelect } from '../../../../../CheckedSelect';
import { PluralDateComponent, pluralDateComponents } from './plural-date-components';
import { NearDateViewProps } from './props';
import { nearDateStyles } from './styles';
import { getThisDateComponentNotation } from './utils';

export const NearDate = withStyles(nearDateStyles)((({
  operands: { shiftAmount, dateComponentType, includeCurrentDate },
  operatorView,
  classes,
  renderBottomSection,
  handleShiftAmountChange,
  handleDateComponentTypeChange,
  handleIncludeCurrentDateChange
}) => {
  return (
    <>
      <div className={classes['near-date']}>
        {operatorView}
        <TextField
          className={classes['text-field']}
          margin='dense'
          variant='outlined'
          value={shiftAmount}
          onChange={handleShiftAmountChange}
          inputProps={{ type: 'number', className: classes['shift-amount'] }}
        />
        <CheckedSelect<PluralDateComponent>
          selectedItem={dateComponentType}
          items={pluralDateComponents}
          onSelect={handleDateComponentTypeChange}
        />
      </div>
      {renderBottomSection(
        <FormControlLabel
          className={classes.FormControlLabel}
          control={
            <Checkbox
              id={'include-date-checkbox'}
              checked={includeCurrentDate}
              onChange={handleIncludeCurrentDateChange}
              color='primary'
              style={{ padding: '0px' }}
            />
          }
          htmlFor={'include-date-checkbox'}
          label={
            <div className={classes.label}>
              Include
              <span className={classes['date-component']}>
                {` ${getThisDateComponentNotation(dateComponentType)}`}
              </span>
            </div>
          }
        />
      )}
    </>
  );
}) as React.FC<NearDateViewProps>);
