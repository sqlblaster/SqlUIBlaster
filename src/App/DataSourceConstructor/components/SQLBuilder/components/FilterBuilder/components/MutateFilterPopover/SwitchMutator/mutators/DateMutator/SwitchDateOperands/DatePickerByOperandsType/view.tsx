import { MenuItem, Select } from '@material-ui/core';
import * as React from 'react';
import { FixedDatePicker } from './components/FixedDatePicker';
import { RelativeDatePicker } from './components/RelativeDatePicker';
import { VariableDatePicker } from './components/VariableDatePicker';
import { dateOperandTypes } from './date-operand-types';
import { DatePickerByOperandsTypeViewProps } from './props';

export const DatePickerByOperandsType: React.FC<
  DatePickerByOperandsTypeViewProps
> = props => {
  const {
    operands,
    limits,
    handleDateTypeChange,
    handleDateChange,
    handleTimeEnabledChange,
    handleRelativeDateChange
  } = props;

  return (
    <>
      <Select
        value={operands.dateType}
        onChange={handleDateTypeChange}
        style={{ marginTop: 10 }}
      >
        {dateOperandTypes.map(dateOperandType => (
          <MenuItem key={dateOperandType} value={dateOperandType}>
            {dateOperandType}
          </MenuItem>
        ))}
      </Select>
      {operands.dateType === 'Fixed date' && (
        <FixedDatePicker
          limits={limits}
          {...operands}
          onDateChange={handleDateChange}
          onTimeEnabledChange={handleTimeEnabledChange}
        />
      )}
      {operands.dateType === 'Relative date' && (
        <RelativeDatePicker
          relativeDate={operands.relativeDate}
          onRelativeDateChange={handleRelativeDateChange}
        />
      )}
      {operands.dateType === 'Variable date' && (
        <VariableDatePicker {...props} />
      )}
    </>
  );
};
