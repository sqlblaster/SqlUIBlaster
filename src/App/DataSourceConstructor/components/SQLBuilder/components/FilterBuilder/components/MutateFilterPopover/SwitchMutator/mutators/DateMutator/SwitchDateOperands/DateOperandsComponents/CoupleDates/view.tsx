import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { DatePickerByOperandsType } from '../../DatePickerByOperandsType';
import { CoupleDateViewProps } from './props';
import { coupleDateStyles } from './styles';

export const CoupleDate = withStyles(coupleDateStyles)((({
  operatorView,
  operands,
  pickedVariables,
  classes,
  handleVariablePick,
  handleOperandsChange
}) => {
  return (
    <>
      <div className={classes.root}>
        {operatorView}
        <div className={classes.dates}>
          <div className={classes.date} data-testid='between-operand-1'>
            <DatePickerByOperandsType
              operands={operands[0]}
              onOperandsChange={handleOperandsChange(0)}
              variable={pickedVariables[0]}
              onVariablePicked={handleVariablePick(0)}
              limits={
                operands[1].dateType === 'Fixed date'
                  ? {
                      maxDate: operands[1].date,
                      maxDateMessage:
                        'Couldn\'t set time, because selected date is later than second date'
                    }
                  : undefined
              }
            />
          </div>
          <div className={classes.date} data-testid='between-operand-2'>
            <DatePickerByOperandsType
              operands={operands[1]}
              onOperandsChange={handleOperandsChange(1)}
              variable={pickedVariables[1]}
              onVariablePicked={handleVariablePick(1)}
              limits={
                operands[0].dateType === 'Fixed date'
                  ? {
                      minDate: operands[0].date,
                      minDateMessage:
                        'Couldn\'t set time, because selected date is earlier than first date'
                    }
                  : undefined
              }
            />
          </div>
        </div>
      </div>
    </>
  );
}) as React.FC<CoupleDateViewProps>);
