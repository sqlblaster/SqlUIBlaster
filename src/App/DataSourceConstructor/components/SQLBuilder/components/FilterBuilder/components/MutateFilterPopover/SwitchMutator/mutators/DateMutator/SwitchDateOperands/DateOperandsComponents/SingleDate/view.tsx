import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { DatePickerByOperandsType } from '../../DatePickerByOperandsType';
import { SingleDateViewProps } from './props';
import { singleDateStyles } from './styles';

// TODO: strech selects to the full available width
export const SingleDate = withStyles(singleDateStyles)((({
  operatorView,
  operands,
  classes,
  pickedVariables,
  handleVariablePicked,
  handleOperandsChange
}) => (
  <div className={classes.root}>
    {operatorView}
    <DatePickerByOperandsType
      operands={operands}
      onOperandsChange={handleOperandsChange}
      variable={pickedVariables}
      onVariablePicked={handleVariablePicked}
    />
  </div>
)) as React.FC<SingleDateViewProps>);
