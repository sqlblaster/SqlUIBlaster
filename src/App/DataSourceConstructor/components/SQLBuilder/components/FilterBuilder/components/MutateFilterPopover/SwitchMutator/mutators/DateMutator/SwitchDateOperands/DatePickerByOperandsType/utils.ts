import { SingleDateOperands } from '../DateOperandsComponents/SingleDate/model';
import { FixedDateOperands } from './components/FixedDatePicker/model';
import { RelativeDateOperands } from './components/RelativeDatePicker/model';
import { VariableDateOperands } from './components/VariableDatePicker/model';
import { DateType } from './date-operand-types';

export const getNewOperands = (dateType: DateType): SingleDateOperands => {
  switch (dateType) {
    case 'Fixed date':
      return new FixedDateOperands();
    case 'Relative date':
      return new RelativeDateOperands();
    case 'Variable date':
      return new VariableDateOperands();
    default:
      throw new Error('unknown date type is provided');
  }
};
