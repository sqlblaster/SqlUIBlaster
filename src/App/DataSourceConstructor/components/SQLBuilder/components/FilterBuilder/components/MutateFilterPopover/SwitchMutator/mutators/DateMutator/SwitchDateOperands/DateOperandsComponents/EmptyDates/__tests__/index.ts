import { fireEvent, getByText as getByTextOn } from 'react-testing-library';
import {
  getGeneratedQuery,
  render
} from 'src/App/DataSourceConstructor/components/SQLBuilder/__tests__/render.utils';
import {
  clickAddTheFilter,
  clickUpdateTheFilter
} from 'src/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/__tests__/utils/filter-utils';
import { selectATable } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/TableSelector/__tests__/utils/select-table';
import {
  getByText$,
  regex
} from 'src/App/DataSourceConstructor/test-utils/utils';
import { clickAddAFilter } from '../../../../../../../../AddFilter/__tests__/utils/click-add-a-filter';
import { mutatorPanelId } from '../../../../../../../view';
import { DateOperator, dateOperators } from '../../../../operators';

describe(`'Is empty' and 'Is not empty' date operations`, () => {
  it(`should render proper sql query,
      should check the proper filter description render
  `, async () => {
    const { getByText, getByTestId, componentNode } = render();

    const columns = selectATable();
    const dateColumnName = columns[2].name;

    const checkFilterDescription = (operator: DateOperator) => {
      expect(
        getByText$(
          componentNode,
          new RegExp(`^${dateColumnName}.*empty.*$`, 'i')
        ).textContent
      ).toMatchSnapshot(`filter description for '${operator}' date operator`);
    };

    clickAddAFilter();
    fireEvent.click(getByText(regex(dateColumnName)));
    fireEvent.click(getByText(regex(dateOperators[0])));
    fireEvent.click(getByText(regex(dateOperators[6])));
    clickAddTheFilter();

    checkFilterDescription('Is empty');

    expect(getGeneratedQuery()).toMatchSnapshot(`'Is empty' operator`);

    fireEvent.click(getByText(regex(dateColumnName)));
    fireEvent.click(
      getByTextOn(getByTestId(mutatorPanelId), regex(dateOperators[6]))
    );
    fireEvent.click(getByText(regex(dateOperators[7])));
    clickUpdateTheFilter();

    checkFilterDescription('Is not empty');

    expect(getGeneratedQuery()).toMatchSnapshot(`'Is not empty' operator`);
  });
});
