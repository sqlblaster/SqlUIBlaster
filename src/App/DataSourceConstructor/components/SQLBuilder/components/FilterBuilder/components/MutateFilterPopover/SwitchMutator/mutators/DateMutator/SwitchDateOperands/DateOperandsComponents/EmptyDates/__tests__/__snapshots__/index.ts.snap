// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'Is empty' and 'Is not empty' date operations should render proper sql query,
      should check the proper filter description render
  : 'Is empty' operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Created At\\" IS NULL
LIMIT 2000
"
`;

exports[`'Is empty' and 'Is not empty' date operations should render proper sql query,
      should check the proper filter description render
  : 'Is not empty' operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Created At\\" IS NOT NULL
LIMIT 2000
"
`;

exports[`'Is empty' and 'Is not empty' date operations should render proper sql query,
      should check the proper filter description render
  : filter description for 'Is empty' date operator 1`] = `"Created Atis empty"`;

exports[`'Is empty' and 'Is not empty' date operations should render proper sql query,
      should check the proper filter description render
  : filter description for 'Is not empty' date operator 1`] = `"Created Atis not empty"`;
